cmake/introspection.cmake

```diff
@@ -0,0 +1,234 @@
+# Copyright (c) 2023-present The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or https://opensource.org/license/mit/.
+
+include(CheckCXXSourceCompiles)
+include(CheckCXXSymbolExists)
+include(CheckIncludeFileCXX)
+
+# The following HAVE_{HEADER}_H variables go to the bitcoin-config.h header.
+check_include_file_cxx(sys/prctl.h HAVE_SYS_PRCTL_H)
+check_include_file_cxx(sys/resources.h HAVE_SYS_RESOURCES_H)
+check_include_file_cxx(sys/vmmeter.h HAVE_SYS_VMMETER_H)
+check_include_file_cxx(vm/vm_param.h HAVE_VM_VM_PARAM_H)
+
+check_cxx_symbol_exists(O_CLOEXEC "fcntl.h" HAVE_O_CLOEXEC)
+
+check_include_file_cxx(unistd.h HAVE_UNISTD_H)
+if(HAVE_UNISTD_H)
+  check_cxx_symbol_exists(fdatasync "unistd.h" HAVE_FDATASYNC)
+  check_cxx_symbol_exists(fork "unistd.h" HAVE_DECL_FORK)
+  check_cxx_symbol_exists(pipe2 "unistd.h" HAVE_DECL_PIPE2)
+  check_cxx_symbol_exists(setsid "unistd.h" HAVE_DECL_SETSID)
+endif()
+
+check_include_file_cxx(sys/types.h HAVE_SYS_TYPES_H)
+check_include_file_cxx(ifaddrs.h HAVE_IFADDRS_H)
+if(HAVE_SYS_TYPES_H AND HAVE_IFADDRS_H)
+  include(TestAppendRequiredLibraries)
+  test_append_socket_library(core_interface)
+endif()
+
+include(TestAppendRequiredLibraries)
+test_append_atomic_library(core_interface)
+
+check_cxx_symbol_exists(std::system "cstdlib" HAVE_STD_SYSTEM)
+check_cxx_symbol_exists(::_wsystem "stdlib.h" HAVE__WSYSTEM)
+if(HAVE_STD_SYSTEM OR HAVE__WSYSTEM)
+  set(HAVE_SYSTEM 1)
+endif()
+
+check_include_file_cxx(string.h HAVE_STRING_H)
```

Comment:
I think we can just drop include checks like this (which don't produce defines used anywhere else in the codebase). I realise this kind of check was ported from Autotools, but I no-longer think it's necessary to do. Running a check to see if we can include `string.h`, just to then immediately compile something which includes `string.h`, seems like it should just be collapsed down to the compile. Looks like another one is `check_include_file_cxx(unistd.h HAVE_UNISTD_H)`.

Reply:
Thanks! Implemented in https://github.com/hebasto/bitcoin/pull/313.

Reply:
Resolved in https://github.com/bitcoin/bitcoin/pull/30454#issuecomment-2281111100.